### PR TITLE
implement new event subscription system

### PR DESCRIPTION
**Pull request by vrabeshko**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -74,18 +74,18 @@
...
-    "eslint-config-airbnb-base": "^13.1.0",
+    "eslint-config-airbnb-base": "^14.1.0",
     "eslint-plugin-import": "^2.13.0",
-    "jest": "^24.8.0",
-    "jest-config": "^24.8.0",
+    "jest": "^25.3.0",
+    "jest-config": "^25.3.0",
     "klaw-sync": "^6.0.0",
-    "lint-staged": "^9.2.0",
+    "lint-staged": "^10.1.0",
```
Comment (dhurlburtusa) [inline package.json:88]: With `lint-staged` 10.x, I think we need to add husky (`npm install --save-dev husky`) after a fresh clone of the repo to make sure the older pre-commit hooks no longer exist.
Then add:
```json
    "husky": {
        "hooks": {
            "pre-commit": "lint-staged"
        }
    },
```
and remove `    "precommit": "lint-staged",`.

File: API.md [hunk 2]
```diff
@@ -387,23 +389,94 @@ standard response object
 - status 400 if error
 
 
-## addEventSubscription(events[, campaignId, config])
+## listEventSubscriptions([config])
+
+list current subscriptions.
```
Comment (dhurlburtusa) [inline API.md:394]:
```suggestion
List current subscriptions.
```

File: API.md [hunk 3]
```diff
@@ -387,23 +389,94 @@ standard response object
...
+- status 200 if success:
+  - result will contain subscription events data
+- status 400 if error
 
-You can register subscriptions to hooks by calling the addEventSubscription function and providing the event (as specified in this document) and handler pairs `eventname: handler`.
 
-There is no wildcard event subscription, however, you can add many handlers at one time.
+## getAvailableSubscriptions([referenceId, referenceType, config])
+
+list available topics and channels to subscribe.
```
Comment (dhurlburtusa) [inline API.md:411]:
```suggestion
List available topics and channels to subscribe.
```

File: API.md [hunk 4]
```diff
@@ -387,23 +389,94 @@ standard response object
...
+
+standard response object
+- status 200 if success:
+  - result will contain subscription topic and channels
+- status 400 if error
+
+
+## getAvailableSubscriptions([config])
+
+list all available topics and channels to subscribe.
```
Comment (dhurlburtusa) [inline API.md:431]:
```suggestion
List all available topics and channels to subscribe.
```

File: API.md [hunk 5]
```diff
@@ -387,23 +389,94 @@ standard response object
...
+
+standard response object
+- status 200 if success:
+  - result will contain subscription topic and channels
+- status 400 if error
+
+
+## addEventSubscriptions(eventSubscriptions[, config])
+
+You can register subscriptions to email, inbox and webhook by calling the addEventSubscriptions function and providing list of event subscription objects.
```
Comment (dhurlburtusa) [inline API.md:448]:
```suggestion
You can register subscriptions to email, inbox, and webhook by calling the `addEventSubscriptions` function and providing list of event subscription objects.
```

File: API.md [hunk 6]
```diff
@@ -425,53 +497,80 @@ standard response object
 - status 400 if error
 
 
-## listEventSubscriptions([campaignId, config])
+## deleteEventSubscriptions(eventSubscriptions, config])
 
-list current subscriptions.
+delete existing subscriptions.
```
Comment (dhurlburtusa) [inline API.md:502]:
```suggestion
Delete existing subscriptions.
```

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 7]
```diff
@@ -1,65 +1,12 @@
...
 
 /**
  *
  * @async
  * @param {string} apiUrlRoot - The root of the API URL.
  * @param {ApiAccessToken} apiAccessToken - A valid API access token.
- * @param {Object} eventsMapping - A mapping between event subscription type and
- *   handler.
- * @param {CampaignId} campaignId - The campaign cursor.
+ * @param {EventSubscription[]} eventSubscriptions - A list of event subscriptions to subscribe
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:9]:
```suggestion
 * @param {EventSubscription[]} eventSubscriptions - A list of event subscriptions to subscribe.

```

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 8]
```diff
@@ -69,47 +16,38 @@ const toEventType = (type) => {
...
+        topic: eventSubscription.topic,
+        channel: eventSubscription.channel,
+        handler: eventSubscription.handler,
+      };
+    }),
   };
   const mutation = `
-    mutation addEventSubscription($config: AddSubscriptionsConfig!) {
-      add_event_subscriptions(config: $config) {
+    mutation addEventSubscriptionss($config: AddSubscriptionsConfig! ){
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:34]:
```suggestion
    mutation addEventSubscriptions($config: AddSubscriptionsConfig!) {

```

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 9]
```diff
@@ -69,47 +16,38 @@ const toEventType = (type) => {
...
+      add_event_subscriptions(config: $config){
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:35]:
```suggestion
      add_event_subscriptions(config: $config) {

```

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 10]
```diff
@@ -151,75 +88,120 @@ const deleteEventSubscription = (apiUrlRoot, apiAccessToken, arrayOfEventNames,
...
-    })
+    .then((response) => response.delete_event_subscriptions)
     .catch((err) => {
       throw ApiUtils.transformError(err);
     });
 };
 
 /**
- * Fetches a page of event subscriptions for the given campaign cursor.
+ * Fetches event subscriptions for the user
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:98]:
```suggestion
 * Fetches event subscriptions for the user.

```

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 11]
```diff
@@ -151,75 +88,120 @@ const deleteEventSubscription = (apiUrlRoot, apiAccessToken, arrayOfEventNames,
...
  *
  * @async
  * @param {string} apiUrlRoot - The root of the API URL.
  * @param {ApiAccessToken} apiAccessToken - A valid API access token.
- * @param {CampaignId} campaignId - The campaign cursor.
+ * @param {string} referenceId - id of related enyity (i.e. campaignId)
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:103]:
```suggestion
 * @param {string} referenceId - ID of related entity (i.e. campaignId).

```

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 12]
```diff
@@ -151,75 +88,120 @@ const deleteEventSubscription = (apiUrlRoot, apiAccessToken, arrayOfEventNames,
...
+ * @param {string} referenceType - type of related entity (i.e. campaign, funding_panel)
  *
  * @returns {Promise<Object>} - The result.
- * @property {EventSubscription[]} data - A page of event subscriptions for the
- *   given campaign cursor.
- * @property {Object} page_info - The page information.
- * @property {boolean} page_info.has_next_page - Flag indicating whether there is
- *   another page of data to may be fetched.
- * @property {string} page_info.end_cursor - The end cursor.
+ * @property {EventSubscription[]} data - A page of event subscriptions for the user
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:107]:
```suggestion
 * @property {EventSubscription[]} data - The event subscriptions for the user.

```

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 13]
```diff
@@ -151,75 +88,120 @@ const deleteEventSubscription = (apiUrlRoot, apiAccessToken, arrayOfEventNames,
...
+    .catch((err) => {
+      throw ApiUtils.transformError(err);
+    });
+};
 
-      // TODO: Remove the following once `event_subscriptions` returns a valid `page_info`
-      // property or no longer does cursor-based pagination.
-      event_subscriptions.page_info = event_subscriptions.page_info || { end_cursor: null, has_next_page: false };
+/**
+ * Fetches available subscriptions and channels for the user
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:137]:
```suggestion
 * Fetches available subscriptions and channels for the user.

```

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 14]
```diff
@@ -151,75 +88,120 @@ const deleteEventSubscription = (apiUrlRoot, apiAccessToken, arrayOfEventNames,
...
+ *
+ * @async
+ * @param {string} apiUrlRoot - The root of the API URL.
+ * @param {ApiAccessToken} apiAccessToken - A valid API access token.
+ * @param {string} referenceId - id of related enyity (i.e. campaignId)
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:142]:
```suggestion
 * @param {string} referenceId - ID of related entity (i.e. campaignId).
```
Please fix any other places this typo was made.

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 15]
```diff
@@ -151,75 +88,120 @@ const deleteEventSubscription = (apiUrlRoot, apiAccessToken, arrayOfEventNames,
...
+ * @param {string} referenceType - type of related entity (i.e. campaign, funding_panel)
+ *
+ * @returns {Promise<Array>} - All available subscriptions to subscribe.
+ *
+ * @throws {XcooBeeError}
+ */
+const getAvailableSubscriptions = (apiUrlRoot, apiAccessToken, referenceId, referenceType) => {
+  const query = `query getAvailableSubscriptions($referenceType: EventReferenceType, $referenceId: String){
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:150]:
```suggestion
  const query = `query getAvailableSubscriptions($referenceType: EventReferenceType, $referenceId: String) {

```

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 16]
```diff
@@ -151,75 +88,120 @@ const deleteEventSubscription = (apiUrlRoot, apiAccessToken, arrayOfEventNames,
...
+    available_subscriptions (reference_type: $referenceType, reference_cursor: $referenceId) {
+        topic
+        channels
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:153]:
```suggestion
      topic
      channels
```
Please fix indentation in other places to 2 spaces too. Thanks.

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 17]
```diff
@@ -151,75 +88,120 @@ const deleteEventSubscription = (apiUrlRoot, apiAccessToken, arrayOfEventNames,
...
+    .then((response) => response.available_subscriptions)
+    .catch((err) => {
+      throw ApiUtils.transformError(err);
+    });
+};
 
-      return event_subscriptions;
-    })
+/**
+ * Fetches available subscriptions and channels for the user
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:168]: Please end sentences with correct punctuation (probably a period in all cases).

File: src/xcoobee/api/EventSubscriptionsApi.js [hunk 18]
```diff
@@ -151,75 +88,120 @@ const deleteEventSubscription = (apiUrlRoot, apiAccessToken, arrayOfEventNames,
...
+ *
+ * @async
+ * @param {string} apiUrlRoot - The root of the API URL.
+ * @param {ApiAccessToken} apiAccessToken - A valid API access token.
+ * @param {string} topic - topic of subscription
+ * @param {string} channel - channel of subscription
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventSubscriptionsApi.js:174]: Please end param descriptions with correct punctuation (probably a period in all cases).
Please start sentences with a capital letter. Thanks.

File: src/xcoobee/api/EventsApi.js [hunk 19]
```diff
@@ -78,29 +83,27 @@ const getEvents = (apiUrlRoot, apiAccessToken, userCursor, privateKey, passphras
...
  * @throws {XcooBeeError}
  */
-const triggerEvent = (apiUrlRoot, apiAccessToken, campaignId, type) => {
+const triggerEvent = (apiUrlRoot, apiAccessToken, topic) => {
   const query = `
-    mutation triggerEvent($campaignId: String!, $type: EventSubscriptionType!) {
-      send_test_event(campaign_cursor: $campaignId, type: $type){
-        event_type
+    mutation sendTestEvent($topic: String){
+      send_test_event(topic: $topic){
```
Comment (dhurlburtusa) [inline src/xcoobee/api/EventsApi.js:96]:
```suggestion
    mutation sendTestEvent($topic: String) {
      send_test_event(topic: $topic) {

```

**Comment by dhurlburtusa** [comment]
@vrabeshko Looks good. Just several minor change requests.

**Comment by dhurlburtusa** [comment]
@Bilal-S Because we are removing functions in this PR, the next release should be a major version update if we are following SemVer. Most developers (I assume) just assume that anything on NPM follows SemVer unless explicitly stated otherwise.

**Comment by vrabeshko** [comment]
@dhurlburtusa applied changes, check pls

**Pull request merged**